compiler/optimizer/src/main/java/edu/snu/nemo/compiler/optimizer/pass/compiletime/annotating/LocationShareAssignmentPass.java

```diff
@@ -0,0 +1,237 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package edu.snu.nemo.compiler.optimizer.pass.compiletime.annotating;
+
+import com.fasterxml.jackson.core.TreeNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import edu.snu.nemo.common.dag.DAG;
+import edu.snu.nemo.common.ir.edge.IREdge;
+import edu.snu.nemo.common.ir.edge.executionproperty.DataCommunicationPatternProperty;
+import edu.snu.nemo.common.ir.vertex.IRVertex;
+import edu.snu.nemo.common.ir.vertex.executionproperty.LocationSharesProperty;
+import edu.snu.nemo.common.ir.vertex.executionproperty.ParallelismProperty;
+import org.apache.commons.math3.optim.BaseOptimizer;
+import org.apache.commons.math3.optim.PointValuePair;
+import org.apache.commons.math3.optim.linear.*;
+import org.apache.commons.math3.optim.nonlinear.scalar.GoalType;
+import org.apache.commons.math3.util.Incrementor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.util.*;
+
+/**
+ * Computes and assigns appropriate share of locations to each stage,
+ * with respect to bandwidth restrictions of locations. If bandwidth information is not given, this pass does nothing.
+ *
+ * <h3>Assumptions</h3>
+ * This pass assumes no skew in input or intermediate data, so that the number of TaskGroups assigned to a location
+ * is proportional to the data size handled by the location.
+ * Also, this pass assumes stages with empty map as {@link LocationSharesProperty} are assigned to locations evenly.
+ * For example, if source splits are not distributed evenly, any source location-aware scheduling policy will
+ * assign TaskGroups unevenly.
+ * Also, this pass assumes network bandwidth to be the bottleneck. Each location should have enough capacity to run
+ * TaskGroups immediately as scheduler attempts to schedule a TaskGroup.
+ */
+public final class LocationShareAssignmentPass extends AnnotatingPass {
+
+  private static final int OBJECTIVE_COEFFICIENT_INDEX = 0;
+  private static final Logger LOG = LoggerFactory.getLogger(LocationShareAssignmentPass.class);
+  private static final HashMap<String, Integer> EMPTY_MAP = new HashMap<>();
+
+  private static String bandwidthSpecificationString = "";
+
+
+  /**
+   * Default constructor.
+   */
+  public LocationShareAssignmentPass() {
+    super(LocationSharesProperty.class, Collections.singleton(ParallelismProperty.class));
+  }
+
+  @Override
+  public DAG<IRVertex, IREdge> apply(final DAG<IRVertex, IREdge> dag) {
+    if (bandwidthSpecificationString.isEmpty()) {
+      dag.topologicalDo(irVertex -> irVertex.setProperty(LocationSharesProperty.of(EMPTY_MAP)));
+    } else {
+      assignLocationShares(dag, BandwidthSpecification.fromJsonString(bandwidthSpecificationString));
+    }
+    return dag;
+  }
+
+  public static void setBandwidthSpecificationString(final String value) {
```

Comment:
When is this method called? 

'bandwidthSpecificationString' appears to determine whether this pass runs or not, which isn't normally the case for optimization pass. Normally, a pass is determined to run when it is included in a policy, and not run when it is not. Maybe this is a good topic for us to discuss offline?

Reply:
It's a temporary replacement for job-wide property.